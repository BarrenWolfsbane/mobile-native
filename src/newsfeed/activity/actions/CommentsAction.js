import React, {
  Component
} from 'react';

import { observer, inject } from 'mobx-react/native';

import {
  TouchableOpacity,
} from 'react-native';

import Icon from 'react-native-vector-icons/MaterialIcons';

import { CommonStyle as CS } from '../../../styles/Common';
import Counter from './Counter';
import withPreventDoubleTap from '../../../common/components/PreventDoubleTap';
import { FLAG_CREATE_COMMENT } from '../../../common/Permissions';

// prevent double tap in touchable
const TouchableOpacityCustom = withPreventDoubleTap(TouchableOpacity);

/**
 * Comments Action Component
 */
export default
@observer
class CommentsAction extends Component {

  static defaultProps = {
    size: 20,
  };
  /**
   * Render
   */
  render() {
    const icon = this.props.entity.allow_comments ? 'chat-bubble' : 'speaker-notes-off';

    const canComment = this.props.entity.allow_comments && this.props.entity.can(FLAG_CREATE_COMMENT);

    const color = canComment ? (this.props.entity['comments:count'] > 0 ? CS.colorPrimary : CS.colorAction) : CS.colorLightGreyed;

    return (
      <TouchableOpacityCustom style={[CS.flexContainer, CS.centered, CS.rowJustifyCenter]} onPress={this.openComments} testID={this.props.testID}>
        <Icon style={[color, CS.marginRight]} name={icon} size={this.props.size} />
        <Counter size={this.props.size * 0.70} count={this.props.entity['comments:count']} />
      </TouchableOpacityCustom>
    );
  }

  /**
   * Open comments screen
   */
  openComments = () => {
    const cantOpen = !this.props.entity.allow_comments && this.props.entity['comments:count'] == 0;
    // TODO: fix
    if (this.props.navigation.state.routeName == 'Activity' || cantOpen) {
      return;
    }
    this.props.navigation.push('Activity', {
      entity: this.props.entity,
      scrollToBottom: true,
    });
  }
}