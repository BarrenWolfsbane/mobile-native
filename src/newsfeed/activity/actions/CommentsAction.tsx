import React, { Component } from 'react';

import { observer } from 'mobx-react';

import { TouchableOpacity } from 'react-native';

import Icon from 'react-native-vector-icons/MaterialIcons';

import { CommonStyle as CS } from '../../../styles/Common';
import Counter from './Counter';
import withPreventDoubleTap from '../../../common/components/PreventDoubleTap';
import { FLAG_CREATE_COMMENT } from '../../../common/Permissions';
import ThemedStyles from '../../../styles/ThemedStyles';
import type ActivityModel from '../../../newsfeed/ActivityModel';
import type BlogModel from '../../../blogs/BlogModel';

// prevent double tap in touchable
const TouchableOpacityCustom = withPreventDoubleTap(TouchableOpacity);

type PropsType = {
  entity: ActivityModel | BlogModel;
  testID?: string;
  size: number;
  navigation: any;
  onPressComment?: () => void;
};

/**
 * Comments Action Component
 */
@observer
class CommentsAction extends Component<PropsType> {
  static defaultProps = {
    size: 20,
  };
  /**
   * Render
   */
  render() {
    const icon = this.props.entity.allow_comments
      ? 'chat-bubble'
      : 'speaker-notes-off';

    const canComment =
      this.props.entity.allow_comments &&
      this.props.entity.can(FLAG_CREATE_COMMENT);

    // const color = canComment
    //   ? this.props.entity['comments:count'] > 0
    //     ? ThemedStyles.style.colorIconActive
    //     : ThemedStyles.style.colorIcon
    //   : CS.colorLightGreyed;

    const color = ThemedStyles.style.colorIcon;

    return (
      <TouchableOpacityCustom
        style={[
          ThemedStyles.style.rowJustifyCenter,
          ThemedStyles.style.paddingHorizontal3x,
          ThemedStyles.style.paddingVertical2x,
          ThemedStyles.style.alignCenter,
        ]}
        onPress={this.openComments}
        testID={this.props.testID}>
        <Icon
          style={[color, CS.marginRight]}
          name={icon}
          size={this.props.size}
        />
        <Counter
          size={this.props.size * 0.7}
          count={this.props.entity['comments:count']}
        />
      </TouchableOpacityCustom>
    );
  }

  /**
   * Open comments screen
   */
  openComments = () => {
    if (this.props.onPressComment) {
      this.props.onPressComment();
      return;
    }
    const cantOpen =
      !this.props.entity.allow_comments &&
      this.props.entity['comments:count'] === 0;
    // TODO: fix
    const routes = this.props.navigation.dangerouslyGetState().routes;
    if ((routes && routes[routes.length - 1].name === 'Activity') || cantOpen) {
      return;
    }
    this.props.navigation.push('Activity', {
      entity: this.props.entity,
      scrollToBottom: true,
    });
  };
}

export default CommentsAction;
