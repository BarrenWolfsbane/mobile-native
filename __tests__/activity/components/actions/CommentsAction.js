import 'react-native';
import React from 'react';
import { Text, TouchableOpacity } from "react-native";
import { shallow } from 'enzyme';

import { activitiesServiceFaker } from '../../../../__mocks__/fake/ActivitiesFaker';

import CommentsAction from '../../../../src/newsfeed/activity/actions/CommentsAction';


describe('Comment action component', () => {

  let screen, navigatorStore, navigation;
  beforeEach(() => {

    const TouchableOpacityCustom = <TouchableOpacity onPress={this.onPress} />;
    navigation = { push: jest.fn(), state: {routeName: 'some'} };
    let activityResponse = activitiesServiceFaker().load(1);
    screen = shallow(
      <CommentsAction entity={activityResponse.activities[0]} navigation={navigation} />
    );

    jest.runAllTimers();
  });

  it('renders correctly', async () => {
    screen.update();
    expect(screen).toMatchSnapshot();
  });


  it('should have a comment button', async () => {

    screen.update();

    expect(screen.find('PreventDoubleTap')).toHaveLength(1)
  });

  it('should navigate a thumb on press ', async () => {

    screen.update();
    let touchables = screen.find('PreventDoubleTap');
    touchables.at(0).props().onPress();
    expect(navigation.push).toHaveBeenCalled();
    expect(screen.find('PreventDoubleTap')).toHaveLength(1);

  });

});
