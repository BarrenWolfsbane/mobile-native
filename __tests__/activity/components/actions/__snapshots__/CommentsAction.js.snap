// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment action component renders correctly 1`] = `
<preventDoubleTap(TouchableOpacity)
  onPress={[Function]}
  style={
    Array [
      undefined,
      undefined,
      undefined,
    ]
  }
>
  <Icon
    allowFontScaling={false}
    name="speaker-notes-off"
    size={20}
    style={
      Array [
        undefined,
        Object {
          "marginRight": 5,
        },
      ]
    }
  />
  <Counter
    size={14}
  />
</preventDoubleTap(TouchableOpacity)>
`;
